### PR TITLE
Synthesis of a Clifford circuit into layers

### PR DESCRIPTION
**Pull request by ShellyGarion**
### Summary
This PR decompose a Clifford into 8 layers of 1-qubit and 2-qubit gates.
There are two layers of CZ gates, and one layer of CX gates.
This PR provides a similar construction to Lemma 8 of https://arxiv.org/abs/2003.09412.

co-authored by @alexanderivrii.

### Details and comments
The aim is to decompose a Clifford for LNN (linear nearest neighbor) connectivity, see #9036 

This PR will be followed by two other PRs for LNN connectivity:

- Decompose a CZ circuit in depth 2n by https://arxiv.org/abs/1705.09176

- Decompose a CX-CZ circuit in depth 5n by https://arxiv.org/abs/2210.16195

Both of these PRs will allow to decompose a Clifford circuit in depth 7n over LNN connectivity.

This code has been used in: https://arxiv.org/abs/2212.03937

### CODE REVIEW

File: qiskit/synthesis/clifford/clifford_decompose_layers.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    q_3: ┤3    ├┤3     ├┤3    ├┤3    ├┤3    ├┤3     ├┤3    ├┤3       ├
+         └─────┘└──────┘└─────┘└─────┘└─────┘└──────┘└─────┘└────────┘
+
+    Args:
+        cliff (Clifford): a clifford operator.
+        cx_synth_func (Callable): a function to decompose the CX sub-circuit.
+        cz_synth_func (Callable): a function to decompose the CZ sub-circuit.
+        cx_cz_synth_func (Callable): optional, a function to decompose the sub-circuits CZ and CX .
+        reverse_cliff (Boolean): if True, reverse the order of the qubits of the Clifford.
+        validate (Boolean): if True, validates the synthesis process.
```
Comment (alexanderivrii) [inline qiskit/synthesis/clifford/clifford_decompose_layers.py:118]: I am actually still confused when we need to reverse the order of the qubits (for the option `reverse_cliff`). Can we remove this option, and instead let the calling application do this reversal when required?

**Review by alexanderivrii — COMMENTED** [review]
Thanks Shelly for driving this development. This is an important decomposition for Clifford circuits that I know many people are eagerly waiting for being available in Qiskit.

I am not sure that I should be the one officially approving this PR, since I am listed as a co-author. I have included a few minor comments, mostly about documentation. 

I am also no longer sure that we need the new `LayeredCircuit` class. Initially it seemed a good idea since the decomposition algorithm produces the final circuit which consists of multiple subcircuits (aka "layers"), including a layer consisting only of Hadamard gates, a layer consisting only of CX-gates. a layer consisting only of CZ-gates, etc. And thus it seemed a good idea that we should first create such a layered decomposition, and then apply layer-specific synthesis algorithms to synthesize each layer (e.g., either Patel-Markov-Hayes or Kutin-Moulton-Smithline linear function synthesis algorithm for CX-layers, a CZ-synthesis algorithm for CZ-layers, etc.), but now we are synthesizing these layers eagerly anyway. Bottom line: let's remove the `LayeredCircuit` class and simply keep a list of QuantumCircuits as you originally had.

Another question is whether we want to use `HighLevelSynthesis` to synthesize layers instead of calling specific `cx_synth_func`, etc.? Oh, I have realized that one does not prevent the other, e.g., we can pass `cx_synth_func` that simply replaces a "CX layer" by `LinearFunction`, and then HLS would be applicable as a follow-up stage. Which means: there is nothing extra needed to do.

File: qiskit/synthesis/clifford/clifford_decompose_layers.py (hunk 1 above)
Comment (alexanderivrii) [reply qiskit/synthesis/clifford/clifford_decompose_layers.py:99]: One more minor comment. Can you also specify the default values for `cx_synth_func` and `cz_synth_func`, i. e. 
```
def synth_clifford_layers(
    cliff,
    cx_synth_func=_default_cx_synth_func,
    cz_synth_func=_default_cz_synth_func,
    cx_cz_synth_func=None,
    reverse_cliff=False,
    validate=False,
):
```

File: qiskit/synthesis/clifford/clifford_decompose_layers.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    Return:
+        QuantumCircuit: a circuit implementation of the Clifford.
+
+    Reference:
+        1. S. Bravyi, D. Maslov, *Hadamard-free circuits expose the
+           structure of the Clifford group*,
+           `arXiv:2003.09412 [quant-ph] <https://arxiv.org/abs/2003.09412>`_
+    """
+
+    cliff_cpy = cliff.copy()
```
Comment (eliarbel) [inline qiskit/synthesis/clifford/clifford_decompose_layers.py:110]: Is the copy here really needed? `cliff_cpy` is only passed to `_create_graph_state` and then it's being copied there as well. So one of those copies seem redundant

File: qiskit/synthesis/clifford/clifford_decompose_layers.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+    CZ2_circ = cz_synth_func(destabz_update, validate=validate)
+
+    mat = destabx.transpose()
+    CX_circ = cx_synth_func(mat, validate=validate)
+
+    return S2_circ, CZ2_circ, CX_circ
+
+
+def _fix_pauli(cliff, cliff_target):
```
Comment (eliarbel) [inline qiskit/synthesis/clifford/clifford_decompose_layers.py:351]: The name of the function suggests a side-effect, but it doesn't affect anything. Maybe call it `_calc_pauli_diff` or something similar

**Review by eliarbel — COMMENTED** [review]
Overall LGTM, with the caveat I didn't get to check the algorithm details... But then I count the unit test to suffice for this purpose. I left a couple minor comments but I don't think those are blockers for merge

**Review by alexanderivrii — APPROVED** [review]
Thanks Shelly! Since I am listed as a co-author on this PR (as I have helped a bit with the initial prototype), I have asked Eli for additional comments. Given that you have also fully addressed his comments, I am marking this as approved.

**Pull request merged**